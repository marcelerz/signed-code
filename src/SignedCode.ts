/**
 * Main class to handle signed code.
 *
 * This class will load a file, validates it, re-signs it, and potentially saves
 * it again.
 * This class is used as the central object for handling signed code.
 *
 * @format
 */

import crypto from "crypto";
import fs from "fs";

const GENERATED_ANNOTATION = "@" + "generated";
const GENERATED_ANNOTATION_REGEX = new RegExp(
	`${GENERATED_ANNOTATION}\\s+(.*?)$`,
	"m",
);
const SIGNED_HASH_ANNOTATION = "@" + "signed-hash";
const SIGNED_HASH_ANNOTATION_REGEX = new RegExp(
	`${SIGNED_HASH_ANNOTATION}\\s+([a-f0-9]+)`,
);
const SIGNED_VERSION_ANNOTATION = "@" + "signed-version";
const SIGNED_VERSION_ANNOTATION_REGEX = new RegExp(
	`${SIGNED_VERSION_ANNOTATION}\\s+([0-9]+)`,
);
const MANUAL_SECTION_INIT = "// <Manual-Section";
const MANUAL_SECTIONS_REGEX = /\/\/\s+<Manual-Section START ([a-zA-Z_][a-zA-Z_0-9-]*?)>(.*?)\/\/\s+<Manual-Section END>/gs;

export const getCommandAnnotation = (command: string): string => {
	return `${GENERATED_ANNOTATION} ${command}`;
};
export const getHashAnnotation = (hash: string): string => {
	return `${SIGNED_HASH_ANNOTATION} ${hash}`;
};
export const getVersionAnnotation = (version: number): string => {
	return `${SIGNED_VERSION_ANNOTATION} ${String(version)}`;
};
export const getManualSection = (name: string, code: string): string => {
	return `${MANUAL_SECTION_INIT} START ${name}>${code}${MANUAL_SECTION_INIT} END>`;
};

const hash = (
	content: string,
	encoding: "utf8" | "ascii" | "latin1",
): string => {
	const md5 = crypto.createHash("md5");
	md5.update(content, encoding);
	return md5.digest("hex");
};

export type ManualSectionInfo = {
	code: string;
};

class SignedCode {
	private content: string;

	public static async load(path: string): Promise<SignedCode> {
		const content = await fs.promises.readFile(path, {
			encoding: "utf8",
			flag: "r",
		});
		return new SignedCode(content);
	}

	constructor(content: string) {
		this.content = content;
	}

	/**
	 * Gets the content given to this instance.
	 */
	getContent(): string {
		return this.content;
	}

	/**
	 * Checks if the hash is valid for the content
	 */
	isValid(): boolean {
		const contentHash = this.getContentHash();
		const derivedHash = this.getDerivedHash();
		return contentHash === derivedHash;
	}

	/**
	 * Determines the hash by looking at the hash given in the content
	 * (non-validated).
	 */
	getContentHash(): string | null {
		const results = SIGNED_HASH_ANNOTATION_REGEX.exec(this.content);
		if (results == null) {
			return null;
		}
		return results[1];
	}

	/**
	 * Determines the hash from its content.
	 */
	getDerivedHash(): string {
		// Do not clean the version. Changes here should trigger a new hash.
		const cleanContent = this.content
			.replace(SIGNED_HASH_ANNOTATION_REGEX, "<hash>")
			.replace(MANUAL_SECTIONS_REGEX, "<manual-code>");
		return hash(cleanContent, "utf8");
	}

	/**
	 * Returns the command string which was used to generate this code.
	 */
	getCommand(): string | null {
		const results = GENERATED_ANNOTATION_REGEX.exec(this.content);
		if (results == null) {
			return null;
		}
		return results[1];
	}

	/**
	 * Returns the version string which was used to generate this code.
	 */
	getVersion(): number | null {
		const results = SIGNED_VERSION_ANNOTATION_REGEX.exec(this.content);
		if (results == null) {
			return null;
		}
		return Number(results[1]);
	}

	/**
	 * Returns all the manual sections found in the code by name.
	 */
	getManualSections(): Record<string, ManualSectionInfo> {
		const regex = new RegExp(MANUAL_SECTIONS_REGEX);
		let captureGroups = null;
		const manualSections: Record<string, ManualSectionInfo> = {};
		do {
			captureGroups = regex.exec(this.content);
			if (captureGroups != null) {
				manualSections[captureGroups[1]] = {
					code: captureGroups[2],
				};
			}
		} while (captureGroups != null);

		return manualSections;
	}

	resign(): string {
		const derivedHash = this.getDerivedHash();
		this.content = this.content.replace(
			SIGNED_HASH_ANNOTATION_REGEX,
			getHashAnnotation(derivedHash),
		);
		return this.content;
	}

	async save(path: string): Promise<void> {
		await fs.promises.writeFile(path, this.content, {
			encoding: "utf8",
			flag: "w",
		});
	}
}

export default SignedCode;
