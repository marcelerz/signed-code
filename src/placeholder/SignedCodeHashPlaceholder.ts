/**
 * Placeholders for signed code.
 *
 * @format
 */

import { getHashAnnotation } from "SignedCode";
import SignedCodePlaceholder from "./SignedCodePlaceholder";

/**
 * Class to handle the hash annotation as placeholder.
 */
export default class SignedCodeHashPlaceholder extends SignedCodePlaceholder {
	constructor() {
		super();
	}

	toString(): string {
		return getHashAnnotation("aa");
	}
}
