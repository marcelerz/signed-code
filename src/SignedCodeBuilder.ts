/**
 * Signed code builder to simplify working with the code generation API.
 *
 * This class is used with the template generation, but can also be used
 * directly.
 *
 * @format
 */

import type { ManualSectionInfo } from "./SignedCode";

import SignedCodePlaceholder from "./placeholder/SignedCodePlaceholder";
import SignedCodeCommandPlaceholder from "./placeholder/SignedCodeCommandPlaceholder";
import SignedCodeHashPlaceholder from "./placeholder/SignedCodeHashPlaceholder";
import SignedCodeManualSectionPlaceholder from "./placeholder/SignedCodeManualSectionPlaceholder";
import SignedCodeVersionPlaceholder from "./placeholder/SignedCodeVersionPlaceholder";
import SignedCode from "./SignedCode";

class SignedCodeBuilder {
	private code: Array<string | SignedCodePlaceholder>;
	private manualSections: Record<string, ManualSectionInfo>;

	constructor() {
		this.code = [];
		this.manualSections = {};
	}

	/**
	 * Add previous code from a previously generated code file.
	 */
	addPreviousCode(code: SignedCode): this {
		this.manualSections = {
			...this.manualSections,
			...code.getManualSections(),
		};
		return this;
	}

	/**
	 * Add a snipet of code.
	 */
	addCode(code: string): this {
		this.code.push(code);
		return this;
	}

	/**
	 * Add a command annotation with the command which defined how the file was
	 * generated.
	 */
	addCommandAnnotation(command: string): this {
		this.code.push(new SignedCodeCommandPlaceholder(command));
		return this;
	}

	/**
	 * Add a hash annotation.
	 */
	addHashAnnotation(): this {
		this.code.push(new SignedCodeHashPlaceholder());
		return this;
	}

	/**
	 * Add a version annotation.
	 */
	addVersionAnnotation(version: number): this {
		this.code.push(new SignedCodeVersionPlaceholder(version));
		return this;
	}

	/**
	 * Generic function to add placeholders.
	 *
	 * Note: This is used by the template engine to add placeholders altogether.
	 */
	addPlaceholder(placeholder: SignedCodePlaceholder): this {
		this.code.push(placeholder);
		return this;
	}

	/**
	 * Sets the code for a manual section by name.
	 *
	 * Note:
	 *   This doesn't add code until a manual section was added to the file
	 *   with the same name.
	 */
	setManualSectionCode(name: string, code: string): this {
		this.manualSections[name] = { code };
		return this;
	}

	/**
	 * Adds a new manual section with name.
	 */
	addManualSection(
		name: string,
		options?: { defaultCode: string; code: string },
	): this {
		this.code.push(
			new SignedCodeManualSectionPlaceholder(
				name,
				options?.code,
				options?.defaultCode,
			),
		);

		return this;
	}

	/**
	 * Get the fully generated code as a signed-code object.
	 */
	getCode(): SignedCode {
		const content = this.code
			.map((entry) => {
				// Set the old value before re-signing
				if (entry instanceof SignedCodeManualSectionPlaceholder) {
					const manualSection = this.manualSections[entry.getName()];
					if (manualSection != null) {
						entry.setPreviousCode(manualSection.code);
					}
				}

				if (entry instanceof SignedCodePlaceholder) {
					return entry.toString();
				}

				return entry;
			})
			.join("");

		const signedContent = new SignedCode(content);
		signedContent.resign();

		return signedContent;
	}

	toString(): string {
		return this.getCode().getContent();
	}
}

export default SignedCodeBuilder;
