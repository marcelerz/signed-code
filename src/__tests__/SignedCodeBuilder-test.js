/**
 * Test for the SignedCodeBuilder module.
 *
 * @format
 */

// SUT
import SignedCodeBuilder from "../SignedCodeBuilder";

import SignedCodeHashPlaceholder from "../placeholder/SignedCodeHashPlaceholder";
import SignedCode from "../SignedCode";

const validChecksum = "824c519dd9ee65cbdd64cc6c5e17be02";
const signedTestCode = (checksum) => `/**
 * This is a test module.
 *
 * @generated /usr/bin/foo -g -lrt
 * @signed-hash ${checksum}
 */

import TestModule from "TestModule";

TestModule.someCall(24, 67);

function printEverything() {
    // <Manual-Section init-code START>
    console.log("42 is the answer!");
    // <Manual-Section END>
}

function someOtherFunction() {
    // <Manual-Section more-code START>
    console.log("I don't know");
    // <Manual-Section END>
}

export default printEverything;
`;

describe("signedCodeBuilder", () => {
	it("accepts plain code", () => {
		const code = "let thereB = 'light';";
		const instance = new SignedCodeBuilder();

		const result = instance.addCode(code);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toStrictEqual(code);
	});

	it("accepts a command annotation", () => {
		const code1 = "let thereA = 'light';\n// ";
		const code2 = "\nlet thereB = 'light';\n";
		const cmd = "/usr/bin/foo -g -lrt";
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addCommandAnnotation(cmd);
		instance.addCode(code2);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});

	it("accepts a hash annotation", () => {
		const code1 = "let thereA = 'light';\n// ";
		const code2 = "\nlet thereB = 'light';\n";
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addHashAnnotation();
		instance.addCode(code2);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});

	it("accepts a version annotation", () => {
		const code1 = "let thereA = 'light';\n// ";
		const code2 = "\nlet thereB = 'light';\n";
		const version = 436;
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addVersionAnnotation(version);
		instance.addCode(code2);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});

	it("accepts a generic placeholder", () => {
		const code1 = "let thereA = 'light';\n// ";
		const code2 = "\nlet thereB = 'light';\n";
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addPlaceholder(new SignedCodeHashPlaceholder());
		instance.addCode(code2);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});

	it("accepts a manual section with code", () => {
		const code1 = "let thereA = 'light';\n";
		const code2 = "\nlet thereB = 'light';\n";
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addManualSection("init-code", {
			code: "let thereC = 'light';",
		});
		instance.addCode(code2);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});

	it("accepts a manual section with default code", () => {
		const code1 = "let thereA = 'light';\n";
		const code2 = "\nlet thereB = 'light';\n";
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addManualSection("init-code", {
			defaultCode: "let thereC = 'light';",
		});
		instance.addCode(code2);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});

	it("accepts a manual section with custom code", () => {
		const code1 = "let thereA = 'light';\n";
		const code2 = "\nlet thereB = 'light';\n";
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addManualSection("init-code", {
			defaultCode: "let thereC = 'light';",
		});
		instance.setManualSectionCode("init-code", "let thereD = 'light';");
		instance.addCode(code2);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});

	it("accepts previous code", () => {
		const previousCode = new SignedCode(signedTestCode(validChecksum));

		const code1 = "let thereA = 'light';\n";
		const code2 = "\nlet thereB = 'light';\n";
		const instance = new SignedCodeBuilder();

		instance.addCode(code1);
		const result = instance.addManualSection("init-code", {
			defaultCode: "let thereC = 'light';",
		});
		instance.setManualSectionCode("init-code", "let thereD = 'light';");
		instance.addCode(code2);
		instance.addPreviousCode(previousCode);

		expect(result).toStrictEqual(instance);
		expect(instance.toString()).toMatchSnapshot();
	});
});
