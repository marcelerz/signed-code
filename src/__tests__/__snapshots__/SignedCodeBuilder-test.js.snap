// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signedCodeBuilder accepts a command annotation 1`] = `
"let thereA = 'light';
// @generated /usr/bin/foo -g -lrt
let thereB = 'light';
"
`;

exports[`signedCodeBuilder accepts a generic placeholder 1`] = `
"let thereA = 'light';
// @signed-hash e1cd25cffbe02a6556a818e034341145
let thereB = 'light';
"
`;

exports[`signedCodeBuilder accepts a hash annotation 1`] = `
"let thereA = 'light';
// @signed-hash e1cd25cffbe02a6556a818e034341145
let thereB = 'light';
"
`;

exports[`signedCodeBuilder accepts a manual section with code 1`] = `
"let thereA = 'light';
// <Manual-Section START init-code>let thereC = 'light';// <Manual-Section END>
let thereB = 'light';
"
`;

exports[`signedCodeBuilder accepts a manual section with custom code 1`] = `
"let thereA = 'light';
// <Manual-Section START init-code>let thereD = 'light';// <Manual-Section END>
let thereB = 'light';
"
`;

exports[`signedCodeBuilder accepts a manual section with default code 1`] = `
"let thereA = 'light';
// <Manual-Section START init-code>let thereC = 'light';// <Manual-Section END>
let thereB = 'light';
"
`;

exports[`signedCodeBuilder accepts a version annotation 1`] = `
"let thereA = 'light';
// @signed-version 436
let thereB = 'light';
"
`;

exports[`signedCodeBuilder accepts previous code 1`] = `
"let thereA = 'light';
// <Manual-Section START init-code>let thereD = 'light';// <Manual-Section END>
let thereB = 'light';
"
`;
